### ISSUE TITLE
Categorical and numerical transformers - sklearn format

### ISSUE THREAD

**Issue by PeterPirog** [ludwig-ai/ludwig#1520]
@w4nderlust @tgaddair I noticed that ludwig is focused mainly on text preprocessing, but there is some weaknes in numerical and categorical encoding.
For example I can't see  possibility to declare by user:

- one-hot encoding, for categorical data with low cardinality datasets (small number of features)
- rare labels encoder, for reducing number of categories for high cardinality datasets
- quantile encoding, for reduce outliers in numerical datasets
- ordinal encoding, for ordered categorical features like features [very small, small, normal, big, very big] =>[0,1,2,3,4]

In my work often I use very good encoders from pages or write own:
[https://feature-engine.readthedocs.io/en/latest/encoding/index.html](https://feature-engine.readthedocs.io/en/latest/encoding/index.html)
[https://contrib.scikit-learn.org/category_encoders/](https://contrib.scikit-learn.org/category_encoders/)
[https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing)

Is there any possibility to  add easily to the project transformers in typical sklearn format, because it will expand ludwig functionality:

```
# These allow us the class to inherit Scikit-learn methods
# such as fit and transform
from sklearn.base import BaseEstimator, TransformerMixin# This function just makes sure that the object is fitted
from sklearn.utils.validation import check_is_fitted

class SubtractMin(BaseEstimator, TransformerMixin):
    def __init__(self, cols_to_operate):
         self.columns = cols_to_operate     
    def fit(self, X, y = None):
         self.min_val_ = X[self.columns].min()
         return self
 
    def transform(self, X):
         # make sure that it was fitted
         check_is_fitted(self, ‘min_val_’)
 
         X = X.copy() # This is so we do not make changes to the            
                        original dataframe
         X[self.columns] = X[self.columns] — self.min_val_
         return X
```
Some improvement can be done by using multivariate imputation (missing values in dataset are predicted with using all features not only single feature)

```
import numpy as np
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
imp_mean = IterativeImputer(random_state=0)
imp_mean.fit([[7, 2, 3], [4, np.nan, 6], [10, 5, 9]])

X = [[np.nan, 2, 3], [4, np.nan, 6], [10, np.nan, 9]]
imp_mean.transform(X)
```

Now I do preprocessing outside the ludwig to prepare csv file but maybe it can be placed inside ludwig in one pipeline.

**Comment by w4nderlust** [comment]
Hi @PeterPirog thanks for this detailed issue!

Yo are right that there are a bunch of additional preprocessing we can do on categorical and numerical features (open to PRs!), but at the same time some of the features you'd like are already here.

- One hot encoding is achieved by setting the category encoder to `sparse`. In a deep learning model is not very useful to be honest, because as soon as that sparse representation is multiplied by a matrix in a fully connected layer, that's equivalent to selecting a row from an embedding matrix, which is what the default `dense` encoder does. So it is there, but it's just a less efficient version of the default option anyway.
- rare labels: there's a `most_common` parameter. If your category feature contains, say 10 distinct values, and you set it to, say, 5, the 5 least common classes will all be encoded with a special token.

Quantile encoding for numerical features would be a great addition, we had it in the backlog for a while. Ordinal encoding on the other hand could be useful but a little bit less so, as it would require the user to specify the mapping manually anyway 8and then they could be doing it in pandas before training with Ludwig), but still a good option to have.

Regarding the common format for transformations, we are going in that direction already with the numerical transformers, but your suggestion is well taken. The main difference with what you are showing would be the assumption that the dataframe operations should work with both pandas and dask.

Regarding imputation, this is also a great suggestion, although the sklearn imputed assumes data in bumpy, and Ludwig doesn't make the same assumption because of the different datatypes. At the same time we could do something potentially similar with pandas/dask.

In general though preprocessing will be a major focus of the v0.6 release, which will start to work on after we are done with the v0.5 PyTorch porting. Really happy to get these detailed suggestions!

**Comment by PeterPirog** [comment]
Hi, @w4nderlust  
I didn't know that some encoders are implemented  (sparse = one-hot). Of course one-hot encoding is completly useless for data with high cardinality so I use Quantile Encoder for categorical data (QuantileEncoder for numerical data and QuantileEncoder for categorical data are completly different)
QuantileEncoding for categorical data is very usefull in spervising learning with very high cardinality:
[https://contrib.scikit-learn.org/category_encoders/quantile.html](https://contrib.scikit-learn.org/category_encoders/quantile.html)
I recommend this article to explain it: [https://maxhalford.github.io/blog/target-encoding/](https://maxhalford.github.io/blog/target-encoding/)

For own purposes I did own transformer for quantile encoding but I didn't describe it yet (code need cleaning):
```
class PercentileTargetEncoder(BaseEstimator, TransformerMixin):
    def __init__(self, features=None,
                 ignored_features=None,
                 p=0.5,
                 m=1,
                 remove_original=True,
                 return_df=True,
                 use_internal_yeo_johnson=True,
                 verbose=True):
        super().__init__()
        self.features = features  # selected categorical features
        self.ignored_features = ignored_features
        self.columns = None  # all columns in df
        self.column_target = None
        self.p = p
        self.m = m
        self.N = None  # Number of rows in training dataset
        self.remove_original = remove_original
        self.return_df = return_df
        # usage of yeo-johnson transformation inside encoder
        self.use_internal_yeo_johnson = use_internal_yeo_johnson
        self.verbose = verbose
        # dict with unique values lists for specified feature, key form (feature)
        self.features_unique = {}
        # stored quantiles for whole dataset, key form (p)
        self.global_quantiles = {}
        # stored quantiles for all values, key form (feature, value, p)
        self.value_quantiles = {}
        # stored counts of every value in train data key form (feature, value)
        self.value_counts = {}

        # convert p and m to lists for iteration available
        if isinstance(p, int) or isinstance(p, float):
            self.p = [self.p]
        if isinstance(m, int) or isinstance(m, float):
            self.m = [self.m]

        # convert feature lists for iteration available
        if not isinstance(self.features, list) and self.features is not None:
            self.features = [self.features]

        if not isinstance(self.ignored_features, list) and self.ignored_features is not None:
            self.ignored_features = [self.ignored_features]

    def fit(self, X, y=None):
        X = X.copy()
        # Convert y to proper datatype
        # if y is pd.Series
        if isinstance(y, pd.Series):
            y = y.to_frame().copy()
        # if y is np.array
        elif isinstance(y, type(np.array([0]))):
            y = pd.DataFrame(y, columns=['target']).copy()
        # if y is pd.DataFrame
        elif isinstance(y, pd.DataFrame):
            y = y.copy()
        else:
            print("Wrong target 'y' data type")

        # use yeo-johnson transformation for target inside encoder
        if self.use_internal_yeo_johnson:
            y = stats.yeojohnson(y)[0]  # second component is lambda
            y = pd.DataFrame(y, columns=['target']).copy()

        # Count number of rows in training dataset
        self.N = len(y)

        self.columns = X.columns
        # Find only categorical columns if not defines
        # Auto-search categorical features
        if self.features is None:
            self.features = [col for col in self.columns if X[col].dtypes == 'O']
        else:
            # convert single feature name to list for iteration possibility
            if isinstance(self.features, str):
                self.features = [self.features]

        # Remove ignored features
        if self.ignored_features is not None:
            for ignored_feature in self.ignored_features:
                self.features.remove(ignored_feature)

        if self.verbose and X.isnull().values.any():
            print('There were some nan values if specified features. Nan values are replaced')

        # Replace nan values in selected categorical features by 'MISSING" value
        # X[self.features] = X[self.features].fillna('MISSING').copy()

        # print (f"self.features={self.features}")

        # Find unique values for specified features
        for feature in self.features:
            self.features_unique[feature] = list(X[feature].unique())
            # print(f"self.features_unique={self.features_unique}")
            # print(f"self.features_unique[feature]={self.features_unique[feature]}")
            # Replace nan values in selected categorical features by 'MISSING" value
            X[feature] = X[feature].apply(lambda value: value if value in self.features_unique[feature] else 'UNKNOWN')

            # add 'UNKNOWN' value for transform never seen values
            self.features_unique[feature].append('UNKNOWN')

            # add 'MISSING' value whole data  were complete and 'MISSING' key is not created
            if not 'MISSING' in self.features_unique[feature]:
                self.features_unique[feature].append('MISSING')

        # Find quantiles for all dataset for each value of p
        for p in self.p:
            self.global_quantiles[p] = np.quantile(y, p)

        # Find quantiles for every feature and every value
        for feature in self.features:
            unique_vals_for_feature = self.features_unique[feature]
            # for every unique value for feature
            for value in unique_vals_for_feature:
                value_counts = X.loc[X[feature] == value, feature].count()

                # value not exist in training data
                if value_counts == 0:
                    for p in self.p:
                        # replace missing value by quantile for all data
                        self.value_quantiles[feature, value, p] = self.global_quantiles[p]

                        # set value 1 for 'UNKNOWN'  and 'MISSING'values
                        self.value_counts[feature, value] = 1

                # value exist in training data, quantile can be calculated
                else:
                    # Find y values for specified feature and specified value
                    idx = X[feature] == value
                    # value_not_exist_in_data=sum(idx.astype(int))
                    y_for_value = y[idx].copy()
                    # counts for every feature and every value
                    self.value_counts[feature, value] = len(y_for_value)

                    for p in self.p:
                        # quantiles calculation
                        quantile = np.quantile(y_for_value, p, interpolation='linear')
                        self.value_quantiles[feature, value, p] = quantile
        return self

    def transform(self, X):
        X = X.copy()
        # Replace nan values in selected categorical features by 'MISSING" value
        # X[self.features] = X[self.features].fillna('MISSING')

        for feature in self.features:
            # Replace never seen values as 'UNKNOWN'
            # X[feature] = X[feature].apply(lambda value: value if value is pd.notnull(value) else 'MISSING')
            X[feature] = X[feature].apply(lambda value: value if pd.notnull(value) else 'MISSING')
            for p in self.p:
                for m in self.m:
                    # Prepare new columns names for percentile values
                    feature_name = feature + '_' + str(p) + '_' + str(m)
                    print(f'feature name={feature_name}')
                    X[feature_name] = X[feature].apply(lambda value:
                                                       self.__calculate_new_target(feature, value, p, m))

        # Remove original features
        if self.remove_original:
            X = X.drop(self.features, axis=1)

        # Return dataframe or np array
        if self.return_df:
            return X
        else:
            return X.to_numpy()

    def __calculate_new_target(self, feature, value, p, m):
        """
        :param feature: current feature name
        :param value: current value for feature
        :param p: percentile value
        :param m: regularization parameter to prevent overfitting , int in range for 1 to np.inf
        :return: calculated target to replace categorical value
        """
        # N - total number of rows in training set
        N = self.N
        # ni - number of rows with specified 'value' in training set
        ni = self.value_counts[feature, value]
        # eta - proportion of specified value to all values in training set
        eta = ni / N
        # q - quantile value for specified 'feature' and specified 'value'
        q = self.value_quantiles[feature, value, p]
        # mQ - mean quantile for whole dataset
        mQ = self.global_quantiles[p]
        # print(f'eth={eta}')
        return (mQ + m * eta * q) / (1 + m * eta)
```
usage:

PercentileTargetEncoder( features=None, # if None detect categorical features or ['feature1,'feature2']
                 ignored_features=None, # do not convert this categorical features None or List 
                 p=0.5,  # percentile - if 0.5 median or list for example [0.1,0.5,0.9]
                 m=1, # smoothing parameter  described in article, in article
                 remove_original=True, #remove original categorical data if True
                 return_df=True, return dataframe if True, nparray if false
                 use_internal_yeo_johnson=True, # inetgrated yeo-yohnson transformation to reduce skewness
                 verbose=True # True to show calculations inside transformer
):

**Comment by w4nderlust** [comment]
An update on this @PeterPirog . We had an internal session, and these features have been prioritized as one of the main features after we are completely done with the PyTorch porting. It will require a little bit more of patience, but they will be added to Ludwig. Thanks for your support and help, very much appreciated!

**Comment by PeterPirog** [comment]
@w4nderlust 
Thank you for the information. Now I test how to effectively encode categorical features with both layers 

-  [ StringLookup layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/StringLookup)
- [Embedding layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding)

I test different vector lenghts for different feature cardinality
